### PR TITLE
Matter zap cleanup

### PR DESCRIPTION
**Pull request by tecimovic**
#### Problem

ZAP source dependency was detrimental to end-customer experience, because every Matter end user had to set themselves up as a ZAP developer: right versions of node, right versions of some shared libraries, etc. The binary build of zap could not be used, because the JS extensions on the Matter SDK reached back into the zap repo looking for source code, which does not exist in a binary build.

#### Change overview

- Allow categorization of helpers
- Move Matter-specific JS extensions into ZAP.
- Externalize the "getResource()" and "collectTests()" mechanisms to support YAML certification files and lists of manual vs. CI tests from JSON files.
- Convert all the templates.json files to stop using helpers on the Matter SDK, but use internal ones.
- Delete all the JS helpers from Matter SDK.
- Allow ZAP_PATH environment variable to point to a different location than third_party/zap/repo for ZAP to use during regeneration.

#### Testing

* zap_regen_all.py was ran repeatedly ensuring that the generation has no difference.
* Other zap scripts (run, convert) were also validated.

#### Follow-up

This work opens up the ability to run zap differently. So there will be follow-up work afterwards:
  - rework the regen script to take full use of multi-core ability
  - rework the regen script to use binary vs source zap interchangeably
  - add the quick-test for the regen to have a quicker more robust CI validation
  - clean up the moved JS code inside zap repo to properly support concurrency.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]
@bzbarsky-apple

**Review by bzbarsky-apple — DISMISSED** [review]
Per Slack discussion, we should land the tests.js changes in a separate PR from the big mass-change, and make sure the certification folks review those bits.  Then rebase this on top of that.

**Comment by tecimovic** [comment]
@bzbarsky-apple : I added a separate PR as we discussed:
   https://github.com/project-chip/connectedhomeip/pull/21784

When that gets merged in to master, I will rebase this PR to consume that, and any other changes from master.

**Review by krypton36 — CHANGES_REQUESTED** [review]
Since deleting items from the templates.json or test.js file, will have no consequence or diff, because the files will remain, but no longer be generated, can you please validate that adding changes to the YAML files and adding tests to the json files are properly generated. Please do not solely rely on the "has no difference" of a ZAP regen.
when validating that the changes are behaving as expected, as CI will not catch removal of items in template or test files as well.

File: examples/chip-tool/templates/tests/commands.zapt [hunk 1]
```diff
@@ -17,7 +17,7 @@ public:
     TestList() : Command("list") {};
     CHIP_ERROR Run() override
     {
-        {{#chip_tests (getTests)}}printf("{{filename}}\n");{{/chip_tests}}
+        {{#chip_tests "../../../../src/app/tests/suites/ciTests.json"}}printf("{{filename}}\n");{{/chip_tests}}
```
Comment (krypton36) [inline examples/chip-tool/templates/tests/commands.zapt:20]: As regards to me previous comment, this change should have caused diffs in zzz_generated, but I do not see them.

**Review by jmartinez-silabs — COMMENTED** [review]
Is the change to submodule silabs/matter_support wanted? I don't see the reason for it

**Comment by tecimovic** [comment]
> Is the change to submodule silabs/matter_support wanted? I don't see the reason for it

@jmartinez-silabs : It's not. I'll remove that change from the PR.

File: examples/chip-tool/templates/tests/commands.zapt (hunk 1 above)
Comment (tecimovic) [reply examples/chip-tool/templates/tests/commands.zapt:20]: @krypton36 : Why should this cause a change? The whole point is that the new system generates exactly same thing. The same data that was earlier returned by `(getTests)` is now taken from the JSON file.

I've done following tests:
   1.) Remove some tests from ciTests and manualTests, I saw the difference in generated stuff.
   2.) Add some tests back into ciTests  and manualTests, I saw the difference in generated stuff.
   
So it seems to me like what you were concerned about above, is working as it should.

**Review by jmartinez-silabs — APPROVED** [review]

File: examples/chip-tool/templates/tests/commands.zapt (hunk 1 above)
Comment (krypton36) [reply examples/chip-tool/templates/tests/commands.zapt:20]: As long as the steps above generated differences, was my main concern. Thank you.

**Review by krypton36 — APPROVED** [review]

File: examples/chip-tool/templates/tests/commands.zapt [hunk 2]
```diff
@@ -48,10 +48,10 @@ void registerCommandsTests(Commands & commands, CredentialIssuerCommands * creds
 #if CONFIG_ENABLE_YAML_TESTS
         make_unique<TestList>(),
         make_unique<ManualTestList>(),
-      {{#chip_tests (getTests)}}
+      {{#chip_tests "../../../../src/app/tests/suites/ciTests.json"}}
```
[the remark this answers is not in the document]
Comment (tecimovic) [reply examples/chip-tool/templates/tests/commands.zapt:51]: Zap doesn't know about CHIP or CHIP root. It only knows about files and directories. While I agree these are ugly, they keep us away from introducing a whole new concept of "SDK" into the zap.

So I would not do this right now. We are slowly introducing the "sdk.json" which might be able to bridge that, but until that becomes mainstream, I would not do that.

**Pull request merged**